service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/aws/storageFolder/ControlledAwsStorageFolderResource.java

```diff
@@ -0,0 +1,213 @@
+    return Optional.of(
+        new UniquenessCheckAttributes()
+            .uniquenessScope(UniquenessScope.GLOBAL)
+            .addParameter("prefix", prefix));
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addCreateSteps(
+      CreateControlledResourceFlight flight,
+      String petSaEmail,
+      AuthenticatedUserRequest userRequest,
+      FlightBeanBag flightBeanBag) {
+    RetryRule cloudRetry = RetryRules.cloud();
+    flight.addStep(
+        new CreateAwsStorageFolderStep(
+            this, flightBeanBag.getAwsCloudContextService(), userRequest),
+        cloudRetry);
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addDeleteSteps(DeleteControlledResourcesFlight flight, FlightBeanBag flightBeanBag) {
+    flight.addStep(
+        new DeleteAwsStorageFolderStep(this, flightBeanBag.getAwsCloudContextService()),
+        RetryRules.cloud());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public void addUpdateSteps(UpdateResourceFlight flight, FlightBeanBag flightBeanBag) {
+    throw new ApiException("addUpdateSteps NotImplemented");
+  }
+
+  public ApiAwsStorageFolderAttributes toApiAttributes() {
+    return new ApiAwsStorageFolderAttributes()
+        .bucketName(getBucketName())
+        .prefix(getPrefix())
+        .region(getRegion());
+  }
+
+  public ApiAwsStorageFolderResource toApiResource() {
+    return new ApiAwsStorageFolderResource()
+        .metadata(super.toApiMetadata())
+        .attributes(toApiAttributes());
+  }
+
+  @Override
+  public String attributesToJson() {
+    return DbSerDes.toJson(
+        new ControlledAwsStorageFolderAttributes(getBucketName(), getPrefix(), getRegion()));
+  }
+
+  @Override
+  public ApiResourceAttributesUnion toApiAttributesUnion() {
+    ApiResourceAttributesUnion union = new ApiResourceAttributesUnion();
+    union.awsStorageFolder(toApiAttributes());
+    return union;
+  }
+
+  @Override
+  public void validate() {
+    super.validate();
+    if (getResourceType() != WsmResourceType.CONTROLLED_AWS_STORAGE_FOLDER
+        || getResourceFamily() != WsmResourceFamily.AWS_STORAGE_FOLDER
+        || getStewardshipType() != StewardshipType.CONTROLLED) {
+      throw new InconsistentFieldsException("Expected CONTROLLED_AWS_STORAGE_FOLDER");
+    }
+    if ((getBucketName() == null) || (getPrefix() == null) || (getRegion() == null)) {
+      throw new MissingRequiredFieldException(
+          "Missing required field for ControlledAwsStorageFolderResource.");
+    }
+  }
+
+  public static class Builder {
+    private ControlledResourceFields common;
+    private String bucketName;
+    private String prefix;
+
+    public Builder common(ControlledResourceFields common) {
+      this.common = common;
+      return this;
+    }
+
+    public Builder bucketName(String bucketName) {
+      this.bucketName = bucketName;
+      return this;
+    }
+
+    public Builder prefix(String prefix) {
```

Comment:
what about just call this folderPath or objectPath

prefix is a little confusing as i don't think the s3 official document uses prefix to refer to s3 object either.

Reply:
Folder is only a visual concept in AWS. It uses prefixes
re: https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html
Furthermore internal operations with AWS SDK work based prefixes, hence prefer to retain prefix